### PR TITLE
Add cuDNN new batch normalization

### PR DESCRIPTION
**Pull request by niboshi**
This PR adds support for new batch normalization interface introduced in cuDNN 7.4.1.

https://docs.nvidia.com/deeplearning/sdk/cudnn-release-notes/rel_741.html

`cupy.cudnn.batch_normalization_forward_training_ex()` returns an extra value "reserve space" that can be passed to the backward API for faster computation.

See also: https://github.com/chainer/chainer/pull/7620

### CODE REVIEW

File: cupy/cudnn.pyx [hunk 1]
```diff
@@ -2042,13 +2098,77 @@ def batch_normalization_forward_training(
...
+
+                reserve_space_size = (
+                    cudnn.getBatchNormalizationTrainingExReserveSpaceSize(
+                        handle,
+                        cudnn_mode,
+                        bn_ops,
+                        0,  # activation desc
+                        x_desc,
+                    ))
+                reserve_space = memory.alloc(reserve_space_size)
```
Comment (emcastillo) [inline cupy/cudnn.pyx]: From the documentation
```
This workspace is not required to be clean. Moreover, the workspace does not have to remain unchanged between the forward and backward pass, as it is not used for passing any information.
```
Maybe we could avoid allocating memory each time if the size remains unchanged?
I think that the overhead should be pretty small, so maybe the current implementation is simpler and better.

Comment (niboshi) [reply cupy/cudnn.pyx]: Yes, CuPy uses memory pool. Allocation is cheap.

**Comment by emcastillo** [comment]
Jenkins, test this please
